File: archived_code/layouts/html_diff.rb
require 'nokogiri/diff'
require 'json'

path = ARGV[0]
archive_html = File.open("#{path}_archive/html.html").read
live_html = File.open("#{path}/html.html").read

live_html = "<div><a><p>AHA</p></a></div>"
archive_html = "<div>fff<p>AHA</p></div>"

archive_html = Nokogiri::HTML(archive_html)
live_html = Nokogiri::HTML(live_html)

changes = []
archive_html.diff(live_html, :added => true) do |change,node|
    # puts node.to_html.ljust(10) + node.parent.path
    changes << {:type => node.class, :text => node.to_html}
  end

opts = {
  array_nl: "\n",
  object_nl: "\n",
  indent: '  ',
  space_before: ' ',
  space: ' '
}
json_str = JSON.generate(changes, opts)
File.write("#{path}_archive/diff.json", json_str)

# doc1 = Nokogiri::HTML('<div><p>one</p> two </div>')
# doc2 = Nokogiri::HTML('<div> two <p>one</p></div>')

# doc1.diff(doc2) do |change,node|
#     puts "AHA"
#     puts node.class
#     # puts "#{change} #{node.to_html}".ljust(30) + node.parent.path
#   end

File: archived_code/layouts/html_diff.py
from xmldiff import main as xmlmain
from xmldiff import formatting as xmlformat
from bs4 import BeautifulSoup
from bs4 import Comment

import json

filter_nodes = ['script'] 

# filter comments and certain nodes
def filter_html(html, filter_nodes):
    soup = BeautifulSoup(html, 'html.parser')
    for node in filter_nodes:
        for tag in soup.find_all(node):
            tag.decompose()
    # decompose comments tag
    for tag in soup.find_all(string=lambda text: isinstance(text, Comment)):
        tag.extract()
    return soup

live_file = f'../recording/pageinfo/test/dimension.html'
archive_file = f'../recording/pageinfo/test_archive/dimension.html'
live_html = open(live_file, 'r').read()
archive_html = open(archive_file, 'r').read()
live_html = str(filter_html(live_html, filter_nodes))
archive_html = str(filter_html(archive_html, filter_nodes))

# live_html = "<div><a><p>AHA</p></a></div>"
# archive_html = "<div>fff<p>AHA</p></div>"

formatter = xmlformat.DiffFormatter(pretty_print=True)
# diff = xmlmain.diff_texts(live_html, archive_html, diff_options={'fast_match':True, "F": 0.99})
diff = xmlmain.diff_texts(live_html, archive_html,
                        diff_options={'F': 0.8, 'ratio_mode': 'accurate'},
                        formatter=formatter)
print(diff)
# pickle.dump(diff, open('diff.pickle', 'wb+'))
# json.dump(diff, open('diff.json', 'w+'), indent=2)

# left = '<document><node>Content</node></document>'
# right = '<document><node>Content</node><a href="/aa">new stuff</a><a href="/aa">new stuff</a></document>'
# diff = xmlmain.diff_texts(left, right)
# print(diff)
# for d in diff:
#     print(type(d))
